src/gazebo_vision_plugin.cpp

```diff
@@ -113,21 +143,178 @@ void VisionPlugin::Load(physics::ModelPtr model, sdf::ElementPtr sdf)
+  ros::SubscribeOptions so_odom = ros::SubscribeOptions::create<nav_msgs::Odometry>(_ros_odom_sub_topic, 1, boost::bind(&VisionPlugin::rosOdomCallBack, this, _1), ros::VoidPtr(), &this->_ros_queue);
+
+  this->_ros_pose_sub = this->_ros_node->subscribe(so_pose);
+  this->_ros_posecov_sub = this->_ros_node->subscribe(so_posecov);
+  this->_ros_odom_sub = this->_ros_node->subscribe(so_odom);
+
+  this->_ros_queue_thread = std::thread(std::bind(&VisionPlugin::queueThread, this));
+#endif
+  } else {
+    // Listen to the update event. This event is broadcast every simulation iteration.
+    _updateConnection = event::Events::ConnectWorldUpdateBegin(
+        boost::bind(&VisionPlugin::OnUpdate, this, _1));
+  }
 
   _pub_odom = _nh->Advertise<nav_msgs::msgs::Odometry>("~/" + _model->GetName() + "/vision_odom", 10);
 }
 
-void VisionPlugin::OnUpdate(const common::UpdateInfo&)
-{
+double VisionPlugin::calcTimeStep() {
 #if GAZEBO_MAJOR_VERSION >= 9
-  common::Time current_time = _world->SimTime();
+  _current_time = _world->SimTime();
 #else
-  common::Time current_time = _world->GetSimTime();
+  _current_time = _world->GetSimTime();
 #endif
-  double dt = (current_time - _last_pub_time).Double();
+  return (_current_time - _last_pub_time).Double();
+}
+
+#if BUILD_ROS_INTERFACE == 1
+void VisionPlugin::handle_vision_data(double sec,
+                                      geometry_msgs::Point pos,
+                                      geometry_msgs::Quaternion att,
+                                      geometry_msgs::Vector3 lin_vel,
+                                      geometry_msgs::Vector3 ang_vel,
+                                      double (&pose_cov)[36],
+                                      double (&twist_cov)[36])
+{
+   // Fill odom msg
+    odom_msg.set_usec(sec * 1e6); // convert sec to usec
+
+    gazebo::msgs::Vector3d* position = new gazebo::msgs::Vector3d();
+    position->set_x(pos.x);
+    position->set_y(pos.y);
+    position->set_z(pos.z);
+    odom_msg.set_allocated_position(position);
+
+    gazebo::msgs::Quaternion* orientation = new gazebo::msgs::Quaternion();
+    orientation->set_x(att.x);
+    orientation->set_y(att.y);
+    orientation->set_z(att.z);
+    orientation->set_w(att.w);
+    odom_msg.set_allocated_orientation(orientation);
+
+    gazebo::msgs::Vector3d* linear_velocity = new gazebo::msgs::Vector3d();
+    linear_velocity->set_x(lin_vel.x);
+    linear_velocity->set_y(lin_vel.y);
+    linear_velocity->set_z(lin_vel.z);
+    odom_msg.set_allocated_linear_velocity(linear_velocity);
+
+    gazebo::msgs::Vector3d* angular_velocity = new gazebo::msgs::Vector3d();
+    angular_velocity->set_x(ang_vel.x);
+    angular_velocity->set_y(ang_vel.y);
+    angular_velocity->set_z(ang_vel.z);
+    odom_msg.set_allocated_angular_velocity(angular_velocity);
+
+    for (int i = 0; i < 36; i++) {
```

Comment:
if you use `std::array`, then you can use `std::copy`.

Reply:
Doesn't work with Gazebo Odometry. If i change it to, for instance:
`std::copy(pose_cov.begin(), pose_cov.end(), odom_msg.pose_covariance.begin());
    std::copy(twist_cov.begin(), twist_cov.end(), odom_msg.twist_covariance.begin()); `
then it gives the error:
`‘((gazebo::VisionPlugin*)this)->gazebo::VisionPlugin::odom_msg.nav_msgs::msgs::Odometry::pose_covariance’ does not have class type
     std::copy(pose_cov.begin(), pose_cov.end(), odom_msg.pose_covariance.begin()); `